sdk/storage/Azure.Storage.Blobs.DataMovement/src/BlobScanner.cs

```diff
@@ -0,0 +1,43 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+using Azure.Storage.Blobs;
+using Azure.Storage.Blobs.Specialized;
+
+namespace Azure.Storage.Blobs.DataMovement
+{
+    /// <summary>
+    /// AzureContainerScanner class.
+    /// </summary>
+    public class BlobScanner
+    {
+        private readonly BlobContainerClient _containerClient;
+
+        /// <summary>
+        /// Constuctor for AzureContainerScanner.
+        /// </summary>
+        /// <param name="client"></param>
+        public BlobScanner(BlobContainerClient client)
+        {
+            // Hacky deep-clone to prevent client changes mid-operation
+            _containerClient = client.GetParentBlobServiceClient().GetBlobContainerClient(client.Name);
+        }
+
+        /// <summary>
+        /// Constuctor for AzureContainerScanner.
+        /// </summary>
+        /// <returns>Async enumerable list containing the name of all blobs in the container.</returns>
+        public async IAsyncEnumerable<string> Scan()
```

Comment:
How result of this is going to be used? Is returning blobName sufficient ?

Reply:
I'm assuming the BlobClients needed to be created for operations will be done by caller using names and the BlobContainerClient it passed in. Is that correct @amnguye ?

Reply:
BlobName/Paths should be enough, since we have all the rest of the info because that was given in order to scan the container.

Reply:
I'm not sure if we should be creating `BlobScanner` just to map BlobItem to Name.

With `System.Linq.Async` nuget package this whole thing can be reduced to :
```
using Azure.Storage.Blobs;
using System.Collections.Generic;
using System.Linq;

namespace ConsoleApp1
{

    class Program
    {
        static void Main(string[] args)
        {
            var containerClient = new BlobContainerClient("REDACTED", "foo");

            IAsyncEnumerable<string> names = containerClient.GetBlobsAsync().Select(x => x.Name);
        }
    }
}
```

I'm not sure if we can take dependency on `System.Linq.Async`. It seems to be used in tests only [here](https://github.com/Azure/azure-sdk-for-net/blob/7397225f873548f4ec317a11099fab0b227454b5/eng/Packages.Data.props#L220). 
However, it shouldn't be hard to convert this class into generic extension like `Select` and put it in internal shared source.

Unless we can find more reasons for `BlobScanner` to exist.

Reply:
I was hoping that the the `BlobScanner` would also scan directories and storage accounts (multiple containers).

But you're right even with that, even with scanning multiple containers, it would probably just be a loop around the `GetBlobsAsync` after we call `ListContainers`

Yeah we can make an internal shared extension to add this helper to.

Reply:
Sounds good. We should bring in Scanner when we face these complexities.

Reply:
Should this go into one of the already existing internal shared classes, or should this go into a new one?

Reply:
Let's keep it in DataMovement for now and make that decision when we see how does it look like.